### DISCUSSION TITLE
Rendering markmap to image in Node environment?

### DISCUSSION THREAD

**Discussion by oddguan** [markmap/markmap#276] [Ideas]
I have some markdown that I would like to render in native iOS/android environment. The best idea that I have right now is to render the mindmap through node.js in the backend and generating an image to be rendered on native mobile environment. I don't need any interaction, just the image. Any better idea to accomplish my goal? Thanks!

**Comment by gera2ld** [comment]
Just found this library using puppeteer and it works like a charm: https://github.com/frinyvonnick/node-html-to-image

```ts
import { Transformer } from 'markmap-lib';
import { fillTemplate } from 'markmap-render';
import nodeHtmlToImage from 'node-html-to-image';
import { writeFile } from 'node:fs/promises';

async function renderMarkmap(markdown: string, outFile: string) {
  const transformer = new Transformer();
  const { root, features } = transformer.transform(markdown);
  const assets = transformer.getUsedAssets(features);
  const html =
    fillTemplate(root, assets, {
      jsonOptions: {
        duration: 0,
      },
    }) +
    `
<style>
body,
#mindmap {
  width: 2400px;
  height: 1800px;
}
</style>
<script>
// This is not needed in the next version.
requestAnimationFrame(() => window.mm.fit());
</script>
`;
  const image = await nodeHtmlToImage({
    html,
  });
  await writeFile(outFile, image);
}

await renderMarkmap(yourMarkdown, 'image.png');
```

**Reply by oddguan** [reply]
Thanks for the suggestion! However, using puppeteer means I need a headless browser environment to do the job, which is pretty hard to scale. Was wondering if there's another way to achieve my goal

**Reply by gera2ld** [reply]
I don't think it's possible because this library requires a browser to render all the HTML elements (converted from Markdown).
